src/MongoDB.Driver.Core/Core/Compression/CompressorFactory.cs

```
@@ -0,0 +1,117 @@
+            _allowedCompressors = Ensure.IsNotNull(allowedCompressors, nameof(allowedCompressors));
+        }
+
+        public ICompressor GetOrCreate(CompressorType compressorType)
+        {
+            if (_cache.TryGetValue(compressorType, out var cachedCompressor))
+            {
+                return cachedCompressor;
+            }
+            else
+            {
+                var requestedCompressor = _allowedCompressors.FirstOrDefault(c => c.Type == compressorType);
+                if (requestedCompressor != null)
+                {
+                    var newCompressor = CreateCompressor(requestedCompressor);
+                    if (newCompressor != null)
+                    {
+                        _cache.Add(compressorType, newCompressor);
+                        return newCompressor;
+                    }
+                }
+
+                if (compressorType == CompressorType.Noop)
+                {
+                    return null;
+                }
+
+                throw new ArgumentException($"The compressor type '{compressorType}' is not allowed.", nameof(compressorType));
+            }
+        }
+
+        private ICompressor CreateCompressor(ConnectionStringCompressorDescription compressorDescription)
+        {
+            switch (compressorDescription.Type)
+            {
+                case CompressorType.Zlib:
+                    {
+                        int? zlibCompressionLevel = null;
+                        if (compressorDescription.Properties.ContainsKey("Level"))
+                        {
+                            zlibCompressionLevel = (int)compressorDescription.Properties["Level"];
+                        }
+
+                        return new ZlibCompressor(zlibCompressionLevel);
+                    }
+                case CompressorType.Snappy:
+                    return new SnappyCompressor();
+            }
+
+            return null;
```

Consider throwing an exception instead of returning null. That way the caller doesn't have to check for null.

---

I will do it